File: src/main/java/ankita/tudubucket/DiscordIntegration.java
package ankita.tudubucket;

import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.FileConfiguration;

import javax.sql.rowset.spi.SyncFactoryException;
import java.io.IOException;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.logging.Logger;

import static org.bukkit.Bukkit.getLogger;

public class DiscordIntegration {

    private final String sendMethod;
    private final String botToken;
    private final String webhookUrl;
    private final String alertChannelId;

    public DiscordIntegration(FileConfiguration config) {
        ConfigurationSection discordConfig = config.getConfigurationSection("integrations.discord");
        assert discordConfig != null;
        sendMethod = discordConfig.getString("method");
        botToken = discordConfig.getString("bot-token");
        webhookUrl = discordConfig.getString("webhook-url");
        alertChannelId = discordConfig.getString("alert-channel-id");
    }

    public void sendMessage(String message) throws IOException, SyncFactoryException {
        try {
            String payload = "{\"content\":\"" + message + "\"}";
            if (!sendMethod.equals("webhook") && !sendMethod.equals("bot")) {
                throw new IllegalStateException("Invalid sending message method! Please check your configuration file again.");
            } else if (sendMethod.equals("bot") && (botToken != null)) {
                // Sending message using a bot
                String url = "https://discord.com/api/channels/" + alertChannelId + "/messages";
                HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Authorization", "Bot " + botToken);
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setDoOutput(true);
                OutputStream outputStream = connection.getOutputStream();
                outputStream.write(payload.getBytes(StandardCharsets.UTF_8));
                outputStream.flush();
                outputStream.close();
                connection.getInputStream().close();
                connection.disconnect();
            } else if (sendMethod.equals("webhook") && (webhookUrl != null)) {
                // Sending message using a webhook
                HttpURLConnection connection = (HttpURLConnection) new URL(webhookUrl).openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setDoOutput(true);
                OutputStream outputStream = connection.getOutputStream();
                outputStream.write(payload.getBytes(StandardCharsets.UTF_8));
                outputStream.flush();
                outputStream.close();
                connection.getInputStream().close();
                connection.disconnect();
            } else {
                // Both bot token and webhook url are not defined
                throw new IllegalStateException("Neither bot token nor webhook url is defined for Discord integration");
            }
        } catch (IOException e) {
            Logger logger = getLogger();
            logger.severe("Failed to send message to Discord: " + e.getMessage());
        }
    }
}

File: src/main/kotlin/net/langdua/wladmin/DiscordIntegration.kt
package net.langdua.wladmin

import net.langdua.bootstrap.PluginConfig
import org.bukkit.Bukkit
import org.bukkit.configuration.ConfigurationSection
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import javax.sql.rowset.spi.SyncFactoryException

class DiscordIntegration(pluginConfig: PluginConfig) {
    private val config = pluginConfig.getConfigurationSection("integrations.discord")?.let { Config.fromConfig(it) }

    class Config(
        val method: String?,
        val token: String?,
        val url: String?,
        val channelId: String?
    ) {
        companion object {
            fun fromConfig(config: ConfigurationSection): Config {
                return Config(
                    config.getString("method"),
                    config.getString("bot-token"),
                    config.getString("webhook-url"),
                    config.getString("alert-channel-id")
                )
            }
        }
    }

    @Throws(IOException::class, SyncFactoryException::class)
    fun sendMessage(message: String) {
        if (config == null) {
            throw IllegalStateException("Improper Discord hook configuration")
        }
        try {
            if ((config.method == "webhook" || config.method == "bot") &&
                (config.method == "bot" && !config.token.isNullOrBlank()) ||
                (config.method == "webhook" && !config.url.isNullOrBlank())
            ) {
                val connection =
                    URL(if (config.method == "bot") "https://discord.com/api/channels/${config.channelId}/messages" else config.url).openConnection() as HttpURLConnection
                if (config.method == "bot") {
                    config.token?.let {
                        connection.setRequestProperty("Authorization", "Bot $it")
                    }
                }
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                val outputStream = connection.outputStream
                val payload =
                    "{\"content\":\"$message\"}" // Both bot token and webhook url are not defined// Sending message using a webhook
                outputStream.write(payload.toByteArray(StandardCharsets.UTF_8))
                outputStream.flush()
                outputStream.close()
                connection.inputStream.close()
                connection.disconnect()
            } else {
                throw IllegalStateException("Neither bot token nor webhook url is defined for Discord integration")
            }
        } catch (e: IOException) {
            val logger = Bukkit.getLogger()
            logger.severe("Failed to send message to Discord: " + e.message)
        }
    }
}
